sharedbuffers/mapped_struct.py

```diff
@@ -297,6 +306,226 @@ def unpack_from(cls, buf, offs, idmap = None, array = array.array, itemsizes = {
+            dataoffs += 8
+            rv = cls(buf, offs, dataoffs, dcode, itemsizes['l'], objlen, idmap)
+        else:
+            raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+        return rv
+
+    @cython.locals(obj_offs = cython.ulonglong, dcode = object, index=cython.ulonglong)
+    def __getitem__(self, index):
+
+        if index >= self.objlen:
+            raise IndexError
+
+        dcode = self.dcode
+        if dcode == 't':
+            if cython.compiled:
+                obj_offs = self.offs + self.index[index]
+            else:
+                index_offs = self.dataoffs + self.itemsize * index
+                obj_offs = self.offs + self.struct.unpack(self.buf[index_offs:index_offs + self.itemsize])[0]
+        else:
+            obj_offs = self.dataoffs + self.itemsize * index
+
+        if  obj_offs in self.idmap:
+            return self.idmap[obj_offs]
+
+        if dcode == 't':
+            res = mapped_object.unpack_from(self.buf, obj_offs, self.idmap)
+        elif cython.compiled:
+            if dcode == 'B':
+                res = cython.cast(cython.p_uchar,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'b':
+                res = cython.cast(cython.p_char,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'H':
+                res = cython.cast(cython.p_ushort,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'h':
+                res = cython.cast(cython.p_short,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'I':
+                res = cython.cast(cython.p_uint,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'i':
+                res = cython.cast(cython.p_int,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'q':
+                res = cython.cast(cython.p_longlong,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            elif dcode == 'd':
+                res = cython.cast(cython.p_double,
+                    cython.cast(cython.p_uchar, self.pybuf.buf) + obj_offs)[0]  # lint:ok
+            else:
+                raise ValueError("Inconsistent data, unknown type code %r" % (dcode,))
+        else:
+            res = self.struct.unpack(self.buf[obj_offs:obj_offs + self.itemsize])[0]
+
+        self.idmap[obj_offs] = res
+        return res
+
+    @cython.locals(a = 'proxied_list')
+    @staticmethod
+    def _eq(a, b):
+        if a.objlen != len(b):
+            return False
+        for i in xrange(a.objlen):
+            if a[i] != b[i]:
+                return False
+        return True
```

Comment:
If it's only for testing that you're implementing `__eq__`, you could simply use generic code here:

``` python
alen = len(a)
if alen != len(b):
    return False
for i in xrange(alen):
    if a[i] != b[i]:
        return False
return True
```

If you want it to be fast, you should have a fast path for the case where both are proxied lists.

You should also not fail in cases where b doesn't support len() or index access (ie: it's another kind of type). You'll need to be safe here if you want to reuse this for mapped tuples (which you probably want), since tuples need to be hashable *and* compare to anything without failing. In fact, lists also compare to anything without failing, so you probably also want that behavior for mapped lists even though I'd agree it'd be best to raise a type error, but python lists don't, so better be consistent with the behavior of the type you're trying to emulate.

You'll also need a full rich compare method (__gt__, etc) for tuples.